src/components/lib/pci/pci.c

```diff
@@ -0,0 +1,131 @@
+/* Simple PCI driver, only works for IVSHMEM device */
+/* PCI background information can be obtained from OSDev http://wiki.osdev.org/PCI */
+/* IVSHMEM device information can be found from qemu documentation */
+
+#include <consts.h>
+#include <assert.h>
+#include "io.h"
+#include "../../../kernel/include/chal.h"
+#include <pci.h>
+#include <llprint.h>
+
+u32_t
+pci_config_read(u32_t bus, u32_t dev, u32_t func, u32_t reg)
+{
+	u32_t v = PCI_ADDR(bus, dev, func, reg);
+	outl(PCI_CONFIG_ADDRESS, v);
+
+	return inl(PCI_CONFIG_DATA);
+}
+
+void
+pci_config_write(u32_t bus, u32_t dev, u32_t func, u32_t reg, u32_t v)
+{
+	u32_t a = PCI_ADDR(bus, dev, func, reg);
+	outl(PCI_CONFIG_ADDRESS, a);
+	outl(PCI_CONFIG_DATA, v);
+}
+
+/* 
+ * iterate through given array sz times to print device's vendor, device id, and classcode
+ */
+void
+pci_dev_print(struct pci_dev *devices, int sz)
+{
+    int i;
+	for(i = 0 ; i < sz ; i++) {
+	 	printc("%x:%x.%x vendor %x device %x class %x\n", devices[i].bus, devices[i].dev, devices[i].func, devices[i].vendor, devices[i].device, devices[i].classcode);
+    }
+}
+
+/* scan through and fill array with devices
+ * return 0 on success
+ * return -1 when sz <= 0
+ */
+int
+pci_scan(struct pci_dev *devices, int sz)
+{
+	int i, j, k, f, tmp, dev_num;
+	u32_t reg;
+	struct pci_bar *bar;
+
+	dev_num = 0;
+
+	if(sz <= 0) return -1;
+
+	for(i = 0 ; i < PCI_BUS_MAX ; i++) {
+		for(j = 0 ; j < PCI_DEVICE_MAX ; j++) {
+			for(f = 0 ; f < PCI_FUNC_MAX ; f++) {
+				reg = pci_config_read(i, j, f, 0x0);
+				if (reg == PCI_BITMASK_32) continue;
+				for(k = 0 ; k < PCI_DATA_NUM ; k++) devices[dev_num].data[k] = pci_config_read(i, j, f, k << 2);
+				devices[dev_num].bus       = (u32_t)i;
+				devices[dev_num].dev       = (u32_t)j;
+				devices[dev_num].func      = (u32_t)f;
+				devices[dev_num].vendor    = (u16_t)PCI_VENDOR_ID(devices[dev_num].data[0]);
+				devices[dev_num].device    = (u16_t)PCI_DEVICE_ID(devices[dev_num].data[0]);
+				devices[dev_num].classcode = (u8_t)PCI_CLASS_ID(devices[dev_num].data[2]);
+				devices[dev_num].subclass  = (u8_t)PCI_SUBCLASS_ID(devices[dev_num].data[2]);
+				devices[dev_num].progIF    = (u8_t)PCI_PROG_IF(devices[dev_num].data[2]);
+				devices[dev_num].header    = (u8_t)PCI_HEADER(devices[dev_num].data[3]);
+				for(k = 0 ; k < PCI_BAR_NUM ; k++) {
+					devices[dev_num].bar[k].raw = devices[dev_num].data[4+k];
+					/* LED: add paddr */
+					devices[dev_num].bar[k].paddr = devices[dev_num].bar[j].raw & 0xFFFFFFF0;
+
+				}
+				
+				dev_num++;
+				if(dev_num >= sz || dev_num >= PCI_DEVICE_MAX) {
+					
+                    return 0;
```

Comment:
The spacing is wonky here. Likely a mismatch between how github renders and your editor? I'm guessing your editor is hiding a tab vs. space mismatch.